### PR TITLE
consensus: add comment as to why use mocks during replay

### LINKED ISSUE #4766: why use the mock.Mempool{} to initialize the blockExec
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): 569981325a43c18e65beb193adf7d172a498a5bc

**What happened**:
* `consensus\replay.go  line 476` : `blockExec := sm.NewBlockExecutor(h.stateDB, h.logger, proxyApp, mock.Mempool{}, sm.MockEvidencePool{})`
* Why use the `mock.Mempool{}` to initialize the blockExec? I notice that the comment showing that  `Mempool` is used for test
* if I add some functions in the interface `Mempool`, the structure `CListMempool` is easy to achieve the additional functions, but the structure `Mempool` is a empty struct. So it is hard for me to achieve the logic code.

**What you expected to happen**:
* use the `CListMempoo`l to initialize the blockExec

### PR DESCRIPTION
**Pull request by melekes**
Closes #4766

### CODE REVIEW

File: consensus/replay.go [hunk 1]
```diff
@@ -474,6 +474,8 @@ func (h *Handshaker) replayBlock(state sm.State, height int64, proxyApp proxy.Ap
 	block := h.store.LoadBlock(height)
 	meta := h.store.LoadBlockMeta(height)
 
+	// Use mocks for both mempool and evidence pool since we don't need them
```
Comment (brapse) [inline consensus/replay.go:477]: Would it be helpful to add an explanation of why we don't need them?

**Comment by erikgrinaker** [comment]
Is it much more expensive to set up an empty mempool and memory-backed evidence pool instead, or simply pass `nil`?

File: consensus/replay.go (hunk 1 above)
Comment (melekes) [reply consensus/replay.go:477]: sure!

**Comment by melekes** [comment]
> Is it much more expensive to set up an empty mempool and memory-backed evidence pool instead, or simply pass nil?

nil does not satisfy the interface. memory backed evidence pool - that's what evmock is!

**Comment by erikgrinaker** [comment]
> > Is it much more expensive to set up an empty mempool and memory-backed evidence pool instead, or simply pass nil?
> 
> nil does not satisfy the interface.

`nil` satisfies all interfaces.

>memory backed evidence pool - we would have to create one. evmock is pretty much the same.

I believe @cmwaters renamed it to `emptyEvidencePool` or something, which I guess is slightly better.

**Comment by melekes** [comment]
> nil satisfies all interfaces.

really?! I will try it then. Is it just panics if you call a method?

**Comment by melekes** [comment]
> nil satisfies all interfaces.

right, I forgot it. Thanks!

**Comment by melekes** [comment]
nil does not work because we call methods on mempoo/evidence pool.

**Comment by erikgrinaker** [comment]
> nil does not work because we call methods on mempoo/evidence pool.

So check if it's nil? If there's a need to run this without mempool/evidence pool, then let's just make them optional?

**Comment by melekes** [comment]
> So check if it's nil? If there's a need to run this without mempool/evidence pool, then let's just make them optional?

too many if conditions?

**Pull request merged**